lib/ruby_lsp/requests/on_type_formatting.rb

```diff
@@ -84,8 +84,30 @@ def handle_statement_end
 
         indents = " " * @indentation
 
-        add_edit_with_text(" \n#{indents}end")
-        move_cursor_to(@position[:line], @indentation + 2)
+        if @previous_line.include?("\n")
+          # If the previous line has a line break, then it means there's content after the line break that triggered
+          # this completion. For these cases, we want to add the `end` after the content and move the cursor back to the
+          # keyword that triggered the completion
+
+          line = @position[:line]
+
+          # If there are enough lines in the document, we want to add the `end` token on the line below the extra
+          # content. Otherwise, we want to insert and extra line break ourselves
+          correction = if T.must(@document.source[@line_end..-1]).count("\n") >= 2
+            line -= 1
+            "#{indents}end"
+          else
+            "#{indents}\nend"
+          end
+
+          add_edit_with_text(correction, { line: @position[:line] + 1, character: @position[:character] })
+          move_cursor_to(line, @indentation + 3)
```

Comment:
Do we want to restore the initial cursor? I would expect the cursor to be: 
``` ruby 
if 
   something
# ^ cursor
```

This reads as behind `end`, so correct me if I'm wrong.

Reply:
The idea of moving the cursor would kinda be to "guess" where the user will type next. If you're breaking the line there, then you'd need to fill in the conditional next. Not sure if it matches everybody's expectation though. What do you think?

Reply:
Oh right! That makes more sense. I was wondering how line was updated to the line for end, but makes more sense that it's at the end of `if `. 

Personally I'd expect to be where I pointed out in the above snippet. My litmus test would be: If I am typing while looking away from the screen, what would me mental cursor be. For me that's after the new line. So I'd expect to hit enter then have to press up to get back to it. 

That said, that could be a bias of not having a tool that is trying to allow for auto complete like that. 

Another case that could be interesting is this one: 
``` ruby 
if (something
```

If the user were to hit enter after the `(` but before the `something`, it would indicate that they are handling a multi-line conditional. 

Though is that handled differently because it's inside `()`?